File: frontend/src/services/api/hotel.ts
import type { ApiListResponse } from "@/types/models/catalog";
import type { ApiResponse } from "@/types/models";
import { apiRequest, qs } from "./http";

export interface HotelSummary {
  room_types: number;
  rooms: number;
  rooms_vacant: number;
  rooms_occupied: number;
  rooms_dirty: number;
  reservations_booked: number;
  in_house: number;
  arrivals_today: number;
  departures_today: number;
  guests: number;
}

export interface HotelRoomType {
  id: string;
  branch_id: string;
  branch_name: string;
  name: string;
  code: string;
  base_rate: number;
  capacity: number;
  description: string;
  is_active: boolean;
  sort_order: number;
}

export interface HotelRoom {
  id: string;
  branch_id: string;
  branch_name: string;
  room_type_id: string;
  room_type_name: string;
  code: string;
  floor: string;
  status: "vacant" | "occupied" | "dirty" | "ooo" | "reserved";
  is_active: boolean;
  notes: string;
}

export interface HotelGuest {
  id: string;
  branch_id: string | null;
  full_name: string;
  phone: string;
  email: string;
  id_number: string;
  notes: string;
  is_active: boolean;
}

export interface FolioLine {
  id: string;
  line_type: string;
  description: string;
  amount: number;
  quantity: number;
  posted_at: string | null;
  notes: string;
}

export interface HotelFolio {
  id: string;
  reservation_id: string;
  branch_id: string;
  status: string;
  balance: number;
  amount_paid?: number;
  outstanding?: number;
  payment_method?: string;
  settled_at?: string | null;
  opened_at: string | null;
  closed_at: string | null;
  notes: string;
  lines?: FolioLine[];
  line_count?: number;
}

export interface HotelReservation {
  id: string;
  reservation_number: string;
  branch_id: string;
  guest_id: string;
  guest_name: string;
  guest_phone: string;
  room_type_id: string;
  room_type_name: string;
  room_id: string | null;
  room_code: string | null;
  status: string;
  check_in_date: string | null;
  check_out_date: string | null;
  nights: number;
  adults: number;
  children: number;
  rate_amount: number;
  notes: string;
  checked_in_at: string | null;
  checked_out_at: string | null;
  folio?: HotelFolio | null;
}

export interface HotelOpenFolio {
  folio_id: string;
  reservation_id: string;
  reservation_number: string;
  room_code: string;
  guest_name: string;
  balance: number;
  branch_id: string;
}

export const hotelApi = {
  summary: (branchId?: string) =>
    apiRequest<ApiResponse<HotelSummary>>(
      `/hotel/summary/${qs({ branch_id: branchId })}`
    ),

  openFolios: (branchId?: string) =>
    apiRequest<ApiResponse<{ results: HotelOpenFolio[]; count: number }>>(
      `/hotel/folios/open/${qs({ branch_id: branchId })}`
    ),
  roomTypes: (page = 1, branchId?: string) =>
    apiRequest<ApiListResponse<HotelRoomType>>(
      `/hotel/room-types/${qs({ page, branch_id: branchId })}`
    ),

  createRoomType: (data: {
    name: string;
    branch_id: string;
    code?: string;
    base_rate?: number;
    capacity?: number;
  }) =>
    apiRequest<ApiResponse<HotelRoomType>>("/hotel/room-types/", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  rooms: (page = 1, branchId?: string, status?: string) =>
    apiRequest<ApiListResponse<HotelRoom>>(
      `/hotel/rooms/${qs({ page, branch_id: branchId, status })}`
    ),

  createRoom: (data: {
    code: string;
    branch_id: string;
    room_type_id: string;
    floor?: string;
  }) =>
    apiRequest<ApiResponse<HotelRoom>>("/hotel/rooms/", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  setRoomStatus: (id: string, status: string) =>
    apiRequest<ApiResponse<HotelRoom>>(`/hotel/rooms/${id}/status/`, {
      method: "POST",
      body: JSON.stringify({ status }),
    }),

  guests: (page = 1, branchId?: string) =>
    apiRequest<ApiListResponse<HotelGuest>>(
      `/hotel/guests/${qs({ page, branch_id: branchId })}`
    ),

  createGuest: (data: {
    full_name: string;
    branch_id?: string;
    phone?: string;
    email?: string;
  }) =>
    apiRequest<ApiResponse<HotelGuest>>("/hotel/guests/", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  reservations: (page = 1, branchId?: string, status?: string) =>
    apiRequest<ApiListResponse<HotelReservation>>(
      `/hotel/reservations/${qs({ page, branch_id: branchId, status })}`
    ),

  createReservation: (data: Record<string, unknown>) =>
    apiRequest<ApiResponse<HotelReservation>>("/hotel/reservations/", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  checkIn: (id: string, roomId?: string) =>
    apiRequest<ApiResponse<HotelReservation>>(`/hotel/reservations/${id}/check-in/`, {
      method: "POST",
      body: JSON.stringify(roomId ? { room_id: roomId } : {}),
    }),

  checkOut: (
    id: string,
    data?: { payment_method?: string; payment_reference?: string }
  ) =>
    apiRequest<ApiResponse<HotelReservation & { settlement?: Record<string, unknown> }>>(
      `/hotel/reservations/${id}/check-out/`,
      {
        method: "POST",
        body: JSON.stringify(data || {}),
      }
    ),

  cancel: (id: string) =>
    apiRequest<ApiResponse<HotelReservation>>(`/hotel/reservations/${id}/cancel/`, {
      method: "POST",
      body: JSON.stringify({}),
    }),

  folio: (id: string) =>
    apiRequest<ApiResponse<HotelFolio>>(`/hotel/reservations/${id}/folio/`),
};

File: backend/apps/platform/demo/hotel.py
"""Hotel demo seeder — room types, rooms, guest, booked + in-house stay."""

from __future__ import annotations

from datetime import timedelta
from decimal import Decimal

from django.utils import timezone

from apps.hotel.models import Reservation, Room
from apps.hotel.services import HotelService
from apps.settings_app.models import Branch
from core.tenancy import tenant_context


def seed(*, tenant, user=None) -> dict:
    with tenant_context(tenant, enforce=True):
        branch = (
            Branch.active_objects()
            .filter(tenant=tenant, is_default=True)
            .first()
            or Branch.active_objects().filter(tenant=tenant).first()
        )
        if branch is None:
            return {"hotel": {"seeded": False, "reason": "no branch"}}

        existing = Room.active_objects().filter(tenant=tenant).count()
        if existing:
            return {
                "hotel": {
                    "seeded": True,
                    "idempotent": True,
                    "rooms": existing,
                    "reservations": Reservation.active_objects()
                    .filter(tenant=tenant)
                    .count(),
                }
            }

        standard = HotelService.create_room_type(
            data={
                "name": "Standard",
                "code": "STD",
                "base_rate": "45.00",
                "capacity": 2,
                "branch_id": branch.id,
                "sort_order": 10,
            },
            user=user,
        )
        deluxe = HotelService.create_room_type(
            data={
                "name": "Deluxe",
                "code": "DLX",
                "base_rate": "75.00",
                "capacity": 3,
                "branch_id": branch.id,
                "sort_order": 20,
            },
            user=user,
        )

        rooms = []
        for code, rtype, floor in (
            ("101", standard, "1"),
            ("102", standard, "1"),
            ("201", deluxe, "2"),
            ("202", deluxe, "2"),
        ):
            rooms.append(
                HotelService.create_room(
                    data={
                        "branch_id": branch.id,
                        "room_type_id": rtype.id,
                        "code": code,
                        "floor": floor,
                    },
                    user=user,
                )
            )

        today = timezone.localdate()
        guest = HotelService.create_guest(
            data={
                "branch_id": branch.id,
                "full_name": "Demo Guest",
                "phone": "+255700000001",
                "email": "demo.guest@example.com",
            },
            user=user,
        )

        booked = HotelService.create_reservation(
            data={
                "branch_id": branch.id,
                "guest_id": guest.id,
                "room_type_id": standard.id,
                "room_id": rooms[1].id,
                "check_in_date": (today + timedelta(days=1)).isoformat(),
                "check_out_date": (today + timedelta(days=3)).isoformat(),
                "adults": 2,
                "rate_amount": "45.00",
                "notes": "Demo arrival tomorrow",
            },
            user=user,
        )

        in_house = HotelService.create_reservation(
            data={
                "branch_id": branch.id,
                "guest_name": "In-House Guest",
                "phone": "+255700000002",
                "room_type_id": deluxe.id,
                "room_id": rooms[2].id,
                "check_in_date": today.isoformat(),
                "check_out_date": (today + timedelta(days=2)).isoformat(),
                "adults": 1,
                "rate_amount": Decimal("75.00"),
            },
            user=user,
        )
        HotelService.check_in(reservation=in_house, user=user)

        return {
            "hotel": {
                "seeded": True,
                "room_types": 2,
                "rooms": len(rooms),
                "guests": 2,
                "reservations": 2,
                "booked": booked.reservation_number,
                "in_house": in_house.reservation_number,
            }
        }
